From edea8c4cd8b651041e8a4aaf6602b707160977f8 Mon Sep 17 00:00:00 2001
From: Cody Peter Mello <cody.mello@joyent.com>
Date: Tue, 16 May 2017 01:46:42 +0000
Subject: [PATCH] OS-5975 VMs inherit inappropriate routes configuration from
 images Reviewed by: Josh Wilsdon <jwilsdon@joyent.com> Approved by: Josh
 Wilsdon <jwilsdon@joyent.com>

---
 src/vm/node_modules/VM.js | 58 ++++++++++++++++++++++++++++-----------
 1 file changed, 42 insertions(+), 16 deletions(-)

diff --git a/src/vm/node_modules/VM.js b/src/vm/node_modules/VM.js
index 153db3cb..3ce23319 100644
--- a/src/vm/node_modules/VM.js
+++ b/src/vm/node_modules/VM.js
@@ -836,6 +836,33 @@ function validateImage(image, log, callback)
     });
 }
 
+function createConfigDir(zonepath, log, cb) {
+    var configDir = path.join(zonepath, 'config');
+
+    /*
+     * Zone images quite often contain the configuration of the zone
+     * they were made from. We remove the old directory here so that
+     * we don't accidentally end up using inherited files.
+     */
+    traceExecFile('/bin/rm', [ '-rf', configDir ],
+        {}, log, 'rm-image-config', function (err) {
+        if (err) {
+            cb(err);
+            return;
+        }
+
+        /*jsl:ignore*/
+        var mode = 0755;
+        /*jsl:end*/
+
+        /*
+         * Now that any pre-existing directory is out of the way, create
+         * one for us to use:
+         */
+        fs.mkdir(configDir, mode, cb);
+    });
+}
+
 // Ensure if image_uuid is passed either at top level or for disks.*.image_uuid
 // that image_uuid exists on the system according to imgadm.
 //
@@ -7355,6 +7382,13 @@ function installZone(payload, log, callback)
             } else {
                 cb();
             }
+        }, function (cb) {
+            if (receiving || reprovisioning) {
+                cb();
+                return;
+            }
+
+            createConfigDir(vmobj.zonepath, log, cb);
         }, function (cb) {
             var createFileOpts = {};
 
@@ -9311,22 +9345,14 @@ exports.reprovision = function (uuid, payload, options, callback)
                 cb(err);
             });
         }, function (cb) {
-            var cmd;
-
-            // copy zones/<uuid>-reprovisioning-root/config to
-            // zones/<uuid>/config so we keep metadata and ipf rules.
-            try {
-                fs.mkdirSync(vmobj.zonepath + '/config');
-            } catch (e) {
-                if (e.code !== 'EEXIST') {
-                    e.message = 'Unable to recreate ' + vmobj.zonepath
-                        + '/config: ' + e.message;
-                    cb(e);
-                    return;
-                }
-            }
-
-            cmd = 'cp -pPR '
+            createConfigDir(vmobj.zonepath, log, cb);
+        }, function (cb) {
+            /*
+             * Copy zones/<uuid>-reprovisioning-root/config to
+             * zones/<uuid>/config so that we keep existing metadata,
+             * IPF rules, routes, and so on.
+             */
+            var cmd = 'cp -pPR '
                 + vmobj.zonepath + '-reprovisioning-root/config/* '
                 + vmobj.zonepath + '/config/';
 
-- 
2.21.0

